File: usim/bus-interface.c
/* bus-interface.c --- CADR bus interface
 *
 */

#include <assert.h>
#include <err.h>
#include <stdbool.h>
#include <stdint.h>

#include "bus-interface.h"
#include "disk-controller.h"
#include "iob.h"
#include "lashup.h"
#include "lashup-debugger.h"
#include "main-memory.h"
#include "tape-controller.h"
#include "tv.h"
#include "ucode.h"
#include "usim.h"
#include "utrace.h"

// "... accumulated error status bits from previous bus cycles."
// "Note that these bits are not cleared by power up."
// only Xbus NXM Error, Unibus NXM Error and Unibus Map Error are used in usim
// others are parity errors, no parity errors in usim
static uint16_t bus_error_status = 0;

// this is controlled by the debugger
static bool nxm_inhibited = false;

// used to detect 1-0 transition of reset bit
static bool modifier_reset = false;

// data = 766100 rw, is not kept because it is the last register accessed
// and it actually executes the operation over socket

// this mirrors the bus_error_status of debuggee
static uint16_t debuggee_bus_error_status;	// 766104 ro

static bool addr17;	// 766110<1> wo															
// reset unibus and bus interface = 766110<2> wo is not kept because it is 
// executed directly
// timeout inhibit = 766110<3> wo is not kept because it is executed directly

static uint16_t addr;	// 766114 wo

static uint32_t
bus_interface_get_debuggee_uaddr()
{
    return ((addr17 ? ((uint32_t)1) : ((uint32_t)0)) << 17) | (((uint32_t)addr) << 1);
}

uint16_t
bus_interface_get_bus_error_status(void)
{
    return bus_error_status;
}

bool
bus_interface_is_xbus_nxm(void)
{
    return (bus_error_status & 01) != 0;
}

bool
bus_interface_is_unibus_nxm(void)
{
    return (bus_error_status & 010) != 0;
}

bool
bus_interface_is_unibus_map_error(void)
{
    return (bus_error_status & 040) != 0;
}

void
bus_interface_reset_bus_error_status(void)
{
    bus_error_status = 0;
}

void
bus_interface_set_debuggee_bus_error_status(uint16_t status)
{
    debuggee_bus_error_status = status;
}

void
bus_interface_set_nxm_inhibit(bool inhibit)
{
    nxm_inhibited = inhibit;
}

void
bus_interface_set_xbus_nxm(void)
{
    if (nxm_inhibited) return;

	// "Xbus NXM Error. 
	// Set when an Xbus cycle times out for lack of response."
	bus_error_status |= 01;
}

// see io/unibus.lisp:UNIBUS-EXISTS-P
void
bus_interface_set_unibus_nxm(void)
{
    if (nxm_inhibited) return;

	// "Unibus NXM Error. 
	// Set when a Unibus cycle times out for lack of response."
	bus_error_status |= 010;
}

void
bus_interface_set_unibus_map_error(void)
{
	// "Unibus Map Error. Set when an attempt to perform an Xbus cycle through
	// the Unibus map is refused because the map specifies invalid or 
	// write-protected."
	bus_error_status |= 040;
}

void
bus_interface_read(uint32_t uaddr, uint16_t *pv)
{
    switch (uaddr) 
	{
        case 0766040:
            *pv = interrupt_status_reg;
			break;

        // 0766042: is write-only

		case 0766044:
            DEBUG(TRACE_BUS_INTERFACE, "bus-interface: read bus status: %06o\n", 
                    bus_error_status);
            *pv = bus_error_status;
			break;

        case 0766100:
            {
                const uint32_t debuggee_uaddr = bus_interface_get_debuggee_uaddr();
                if (lashup_debugger_read(debuggee_uaddr, pv))
                {
                    DEBUG(TRACE_BUS_INTERFACE, "bus-interface: read data: %06o from %06o\n", 
                            *pv, uaddr);
                }
                else
                {
                    WARNING(TRACE_USIM, "bus-interface: read failed\n");
                    *pv = 0;
                }
            }
            break;

        // status for bus cycles executed on the debuggee's busses
        // read only
        // cleared by writing into debuggee's 766044 (error status)
        case 0766104:
            *pv = debuggee_bus_error_status;
            DEBUG(TRACE_BUS_INTERFACE, "bus-interface: read status: %06o\n", *pv);
            break;

		default:
			warnx("bus-interface: read invalid uaddr:%6o", uaddr);
            bus_interface_set_unibus_nxm();
	}

}

void
bus_interface_write(uint32_t uaddr, uint16_t v)
{
    switch (uaddr) 
	{
        case 0766040:
            // "Writing this location writes into bits 0 and 10-13 
            // (mask 36001) of the above register. This is used to 
            // change the "interrupt level" and to re-enable acceptance 
            // of Unibus interrupts after processing an interrupt."
            set_interrupt_status_reg((interrupt_status_reg & ~0036001) | (v & 0036001));
			break;

		case 0766042:
            // "Writing this location writes into bits 2-9 and 15 
            // (mask 101774) of the above register. This is used to simulate 
            // Unibus interrupts and to clear bit 15 (Unibus Interrupt) 
            // after processing an interrupt."
            set_interrupt_status_reg((interrupt_status_reg & ~0101774) | (v & 0101774));
			break;

		case 0766044:
            DEBUG(TRACE_BUS_INTERFACE, "bus-interface: write (clear) bus status\n");
            // "Writing this location ignores the data written and 
            // clears the status bits."
            bus_error_status = 0;

            // "766104 -> These bits are cleared by writing into location 766044
            // (Error Status) on the debuggee's Unibus."
            debuggee_bus_error_status = 0;
			break;

        case 0766100:
            {
                const uint32_t debuggee_uaddr = bus_interface_get_debuggee_uaddr();
                DEBUG(TRACE_BUS_INTERFACE, "bus-interface: write data: %06o to %06o\n", v, uaddr);
                if (lashup_debugger_write(debuggee_uaddr, v))
                {
                    // writing into debuggee's error status (766044)
                    // resets debugger's status also
                    if (debuggee_uaddr == 0766044)
                    {
                        debuggee_bus_error_status = 0;
                    }
                }
            }
            break;

        case 0766102:
            {
                const uint8_t cmd = ((v >> 8) & 0377);
                const uint8_t param = (v & 0377);

                DEBUG(TRACE_BUS_INTERFACE, "bus-interface: remote usim command: #o%o, param: #o%o\n");

                lashup_debugger_usim(cmd, param);
            }
            break;
        
        case 0766110:
            {
                addr17 = ((v & 01) != 0);
                const bool reset = ((v & 02) != 0);
                const bool timeout_inhibit = ((v & 04) != 0);
                const bool debuggee_mark = ((v & 010) != 0);
                const bool debugger_mark = ((v & 020) != 0);
                const bool ping = ((v & 040) != 0);
                const uint8_t mark_symbol = ((v >> 8) & 0377);
                DEBUG(TRACE_BUS_INTERFACE, "bus-interface: write modifier bits %06o: addr17:%d reset:%d timeout_inhibit:%d\n", 
                        v, addr17, reset, timeout_inhibit);

                if (debugger_mark) lashup_debugger_mark_debugger(mark_symbol); 
                if (debuggee_mark) lashup_debugger_mark_debuggee(mark_symbol);
                // break so timeout request is not sent
                if (debugger_mark || debuggee_mark) break;

                if (ping)
                {
                    lashup_debugger_ping();
                    break;
                }

                if ((modifier_reset) && (!reset))
                {
                    if (!lashup_debugger_reset_unibus_and_bus_interface())
                    {
                        DEBUG(TRACE_BUS_INTERFACE, "bus-interface: cannot execute reset\n");
                    }
                }
                modifier_reset = reset;
                uint16_t timeout_inhibit_data = timeout_inhibit ? 1 : 0; 
                if (!lashup_debugger_inhibit_nxm(timeout_inhibit_data))
                {
                    DEBUG(TRACE_BUS_INTERFACE, "bus-interface: cannot execute timeout inhibit\n");
                }
            }
            break;

        case 0766112:
            {
                __attribute__((unused)) const uint8_t cmd = ((v >> 8) & 0377);
                __attribute__((unused)) const uint8_t param = (v & 0377);
                DEBUG(TRACE_BUS_INTERFACE, "bus-interface: local usim command: #o%o, param: #o%o\n",
                        cmd, param);
            }
            break;


        case 0766114:
            addr = v;
            DEBUG(TRACE_BUS_INTERFACE, "bus-interface: write address: %06o (uaddr:%6o)\n", 
                    v, bus_interface_get_debuggee_uaddr());
            break;

		default:
			warnx("bus-interface: write invalid uaddr:%6o", uaddr);
            bus_interface_set_unibus_nxm();
	}

}

void 
bus_interface_bus_reset(void)
{
    bus_error_status = 0;
    nxm_inhibited = false;
    modifier_reset = false;
    debuggee_bus_error_status = 0;
    addr17 = 0;
    addr = 0;

    // Unibus reset (BUS.INIT L)
    iob_bus_reset();
    tape_controller_bus_reset();
    
    // Xbus reset (XBUS.INIT L)
    main_memory_bus_reset();
    disk_controller_bus_reset();
    tv_bus_reset();
}

File: usim-cs/BusInterface.cs
// BusInterface.cs - Faithful port of usim/bus-interface.c's bus-error-status
// and interrupt-register block. Deliberately excludes the "lashup" remote-
// debugger protocol (a serial/socket link to a SECOND, PHYSICAL CADR machine
// acting as a debuggee) -- a software-only emulator has no second machine to
// talk to, so every lashup-only register becomes a logged no-op (or, for
// 0766104, a faithfully-always-0 read: with no debuggee ever attached, 0 IS
// the real C's correct answer, not a compromise). See
// docs/superpowers/specs/2026-09-07-bus-interface-design.md for the full
// register-by-register rationale.

using System;

namespace Usim;

public class BusInterface
{
    private readonly UCode _ucode;

    // Mirrors the real C's file-scope statics (usim/bus-interface.c:26-45).
    // modifier_reset and debuggee_bus_error_status are lashup-only state --
    // omitted, since every call site that would read or write them is
    // already a no-op in this port.
    private ushort _busErrorStatus;
    private bool _nxmInhibited;
    private bool _addr17;
    private ushort _addr;

    public BusInterface(UCode ucode)
    {
        _ucode = ucode;
    }

    // usim/bus-interface.c:53-57
    public ushort GetBusErrorStatus() => _busErrorStatus;

    // usim/bus-interface.c:59-75. Masks independently re-verified via
    // script: Xbus NXM = octal 01 = 0x1, Unibus NXM = octal 010 = 0x8,
    // Unibus Map Error = octal 040 = 0x20.
    public bool IsXbusNxm() => (_busErrorStatus & 0x1) != 0;
    public bool IsUnibusNxm() => (_busErrorStatus & 0x8) != 0;
    public bool IsUnibusMapError() => (_busErrorStatus & 0x20) != 0;

    // usim/bus-interface.c:77-81
    public void ResetBusErrorStatus() => _busErrorStatus = 0;

    // usim/bus-interface.c:89-93
    public void SetNxmInhibit(bool inhibit) => _nxmInhibited = inhibit;

    // usim/bus-interface.c:95-103. Gated by nxm_inhibited.
    public void SetXbusNxm()
    {
        if (_nxmInhibited) return;
        _busErrorStatus |= 0x1;
    }

    // usim/bus-interface.c:106-114. Gated by nxm_inhibited.
    public void SetUnibusNxm()
    {
        if (_nxmInhibited) return;
        _busErrorStatus |= 0x8;
    }

    // usim/bus-interface.c:116-123. NOT gated by nxm_inhibited -- this
    // asymmetry with SetXbusNxm/SetUnibusNxm is in the real C as written.
    public void SetUnibusMapError()
    {
        _busErrorStatus |= 0x20;
    }

    /// <summary>
    /// Faithful port of bus_interface_read (usim/bus-interface.c:125-171),
    /// minus lashup. uaddr is the Unibus word address, matching BusAdaptor's
    /// ReadUnibus/WriteUnibus convention.
    /// </summary>
    public uint Read(uint uaddr)
    {
        switch (uaddr)
        {
            case 0x3EC20: // 0766040 octal
                return (uint)_ucode.InterruptStatusReg;

            // 0766042 is write-only in the real C -- falls to default below.

            case 0x3EC24: // 0766044 octal
                TraceLog.Instance.Debug(TraceCategory.Memory,
                    $"bus-interface: read bus status: {Convert.ToString(_busErrorStatus, 8)}");
                return _busErrorStatus;

            case 0x3EC40: // 0766100 octal
                // Real C attempts a lashup read of the debuggee's bus; no
                // debuggee exists here.
                TraceLog.Instance.Warning(TraceCategory.Memory,
                    "bus-interface: read data -- no debuggee attached (lashup not ported)");
                return 0;

            case 0x3EC44: // 0766104 octal
                // Debuggee's mirrored bus-error status (lashup-only). Always
                // 0 here -- the real, correct answer with no debuggee ever
                // attached, not a placeholder.
                return 0;

            default:
                TraceLog.Instance.Warning(TraceCategory.Memory,
                    $"bus-interface: read invalid uaddr:{Convert.ToString((int)uaddr, 8)}");
                SetUnibusNxm();
                return 0;
        }
    }

    /// <summary>
    /// Faithful port of bus_interface_write (usim/bus-interface.c:173-292),
    /// minus lashup.
    /// </summary>
    public void Write(uint uaddr, uint v)
    {
        switch (uaddr)
        {
            case 0x3EC20: // 0766040 octal
                // "Writing this location writes into bits 0 and 10-13 (mask
                // 36001)." Octal 036001 = 0x3C01 -- independently verified
                // via script; a naive hand-conversion gives the wrong
                // 0xF001 (bits 0, 12-15).
                _ucode.SetInterruptStatusReg((_ucode.InterruptStatusReg & ~0x3C01) | ((int)v & 0x3C01));
                break;

            case 0x3EC22: // 0766042 octal
                // "Writing this location writes into bits 2-9 and 15 (mask
                // 101774)." Octal 0101774 = 0x83FC.
                _ucode.SetInterruptStatusReg((_ucode.InterruptStatusReg & ~0x83FC) | ((int)v & 0x83FC));
                break;

            case 0x3EC24: // 0766044 octal
                TraceLog.Instance.Debug(TraceCategory.Memory,
                    "bus-interface: write (clear) bus status");
                _busErrorStatus = 0;
                break;

            case 0x3EC40: // 0766100 octal
                // Real C writes to the debuggee's bus over lashup; no-op here.
                TraceLog.Instance.Warning(TraceCategory.Memory,
                    $"bus-interface: write data 0x{v:X} -- no debuggee attached (lashup not ported)");
                break;

            case 0x3EC42: // 0766102 octal
                // Remote usim command over lashup; no-op here.
                TraceLog.Instance.Warning(TraceCategory.Memory,
                    $"bus-interface: remote usim command 0x{v:X} -- no debuggee attached (lashup not ported)");
                break;

            case 0x3EC48: // 0766110 octal
                // Modifier bits. Only addr17 (bit 0) has any meaning without
                // lashup (it's part of the debuggee-target-address
                // computation, itself only consumed by the 0766100 lashup
                // path) -- captured anyway as harmless bookkeeping. Every
                // other bit (reset, timeout-inhibit, debugger/debuggee mark,
                // ping) drives a lashup_debugger_* call in the real C; all
                // become logged no-ops here.
                _addr17 = (v & 0x1) != 0;
                TraceLog.Instance.Debug(TraceCategory.Memory,
                    $"bus-interface: write modifier bits 0x{v:X} (addr17={_addr17}) -- reset/timeout-inhibit/mark/ping are lashup-only, no-op");
                break;

            case 0x3EC4A: // 0766112 octal
                // Local usim command -- already a log-only no-op in the real
                // C itself (its cmd/param params are marked unused there).
                TraceLog.Instance.Debug(TraceCategory.Memory,
                    $"bus-interface: local usim command 0x{v:X}");
                break;

            case 0x3EC4C: // 0766114 octal
                _addr = (ushort)v;
                TraceLog.Instance.Debug(TraceCategory.Memory,
                    $"bus-interface: write address: 0x{v:X}");
                break;

            default:
                TraceLog.Instance.Warning(TraceCategory.Memory,
                    $"bus-interface: write invalid uaddr:{Convert.ToString((int)uaddr, 8)}");
                SetUnibusNxm();
                break;
        }
    }

    /// <summary>
    /// Faithful port of bus_interface_bus_reset (usim/bus-interface.c:294-312),
    /// minus lashup state. The real C's fan-out to iob_bus_reset()/
    /// main_memory_bus_reset()/disk_controller_bus_reset() is a genuine no-op
    /// even there (all three are empty function bodies) -- nothing to call.
    /// tape_controller_bus_reset()/tv_bus_reset() would do something (clear a
    /// status struct; the latter is empty too) but neither TapeController nor
    /// TV exist in C# yet -- logged as a placeholder note, not silently
    /// dropped.
    /// </summary>
    public void BusReset()
    {
        _busErrorStatus = 0;
        _nxmInhibited = false;
        _addr17 = false;
        _addr = 0;

        TraceLog.Instance.Info(TraceCategory.Memory,
            "bus-interface: bus reset (tape-controller/tv bus_reset not yet ported -- no-op)");
    }
}
